Refactor messaging attributes and per-message attributes in batching scenarios (#2763)
open-telemetry/opentelemetry-specification: specification/trace/semantic_conventions/messaging.md

## Messaging attributes

> Following messaging attributes apply to spans describing `send`, `receive`, and `process`  operations.

Author: removed the sentence, as it does not clarify anything.

| [`net.peer.name`](span-general.md) | string | Logical remote hostname, see note below. [9] | `example.com` | Conditionally Required: If available. |
| [`net.sock.family`](span-general.md) | string | Protocol [address family](https://man7.org/linux/man-pages/man7/address_families.7.html) which is used for communication. | `inet6`; `bluetooth` | Conditionally Required: [10] |
| [`net.sock.peer.addr`](span-general.md) | string | Remote socket peer address: IPv4 or IPv6 for internet protocols, path for local communication, [etc](https://man7.org/linux/man-pages/man7/address_families.7.html). | `127.0.0.1`; `/tmp/mysql.sock` | Recommended |
| [`net.sock.peer.name`](span-general.md) | string | Remote socket peer name. | `proxy.example.com` | Recommended: [11] |
| [`net.sock.peer.port`](span-general.md) | int | Remote socket peer port. | `16456` | Recommended: [12] |

**[1]:** If the message destination is either a `queue` or `topic`.

**[2]:** If value is `true`. When missing, the value is assumed to be `false`.

> **[3]:** Instrumentation SHOULD always set it on batch `receive` operations regardless of the actual number of received messages (e.g. 0, 1, or more).

Reviewer: Is this note needed? I feel like note 4 below covers this. Perhaps if the desire is to express "SHOULD", this could be added at the beginning of note 4?

Author: This note describes `receive` specifically, while  note above applies to all possible messaging operations.
Receive is special because you can have 0 messages received and also it's common to request N messages and potentially receive M, so defaulting to 1 does not make sense. 

This is also a recommendation, while `conditionally_required` level on note [4] implies MUST.

> **[4]:** If available and only if the span describes operations that operate with message batches.

Reviewer: ```suggestion
**[4]:** If available within the messaging system, and only if the span describes operations that operate with message batches.
```
The reason for this change is that before it looked like the intent was to say "Available if and only if", except a mistake was made. I think this makes it clearer.

> **[5]:** only if messaging.batch.size is not set or set to `1`.

Reviewer: ```suggestion
**[5]:** Only if messaging.batch.size is not set or if the value applies to all messages in the batch.
```
The purpose for this suggestion is to leave the door open to the possibility all messages in a batch could be part of the same conversation. Regardless of whether any existing messaging systems behave this way, I see no reason to insist the batch size needs to be 1 for this to apply to a batch.

**[6]:** only if messaging.batch.size is not set or set to `1`.

**[7]:** only if messaging.batch.size is not set or set to `1`.

**[8]:** only if messaging.batch.size is not set or set to `1`.

> 

Reviewer: Can note 5 be reused for all of these?

Author: The source of truth is messaging.yaml file, this notes are auto-generated and I feel it's important to have conditions specified on each attribute.

Furthermore, it is strongly recommended to add the [`net.transport`][] attribute and follow its guidelines, especially for in-process queueing systems (like [Hangfire][], for example).
These attributes should be set to the broker to which the message is sent/from which it is received.

> Note that attributes in `messaging.message` namespace describe message. It's recommended to use `messaging.{system}.message` namespace for specific messaging system attributes that describe individual messages.

Reviewer 2: >namespace describe message

I'm a bit confused with this sentence. Should it also say?

>namespace describe an individual message.

Author: Discussed at SIG meeting:
- `messaging.{system}.message.attr` is preferable to keep all system-specific attr in the same namespace
- `{system}` should match exactly `messaging.system`
- we should also describe namespace pattern for messaging.destination and source.
- Perhaps we should standardize it across specs in https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-naming.md

Author: * allow jms and amqp and specific system on the span
* messaging.system vs messaging.{system}.message needs further discussion, but keep it open for now.

Reviewer 3: Sorry, I missed the discussion about multiple messaging system combinations last week. My take on this:
* AMQP qualifies more than a protocol than a system. So, you might have a `messaging.system` set to `kafka`, and `net.app.protocol.name` set to `amqp`.
* There's the case when AMQP is used without any messaging system on top. I think at one point there should be semantic conventions specific to AMQP that (probably) prescribe to dedicated spans for AMQP operations and can be combined with spans prescribed by these conventions. AMQP semantic conventions would model the transport layer.
* I think if messaging systems are used on top of each other (e. g. JMS and RabbitMQ) each of those will create spans independently. This causes other problems (@lmolkova treats them in https://github.com/open-telemetry/oteps/pull/172), however, I think it's the only reasonable way to model this reliable.

| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|
| `messaging.system` | string | A string identifying the messaging system. | `kafka`; `rabbitmq`; `rocketmq`; `activemq`; `AmazonSQS` | Required |
| `messaging.operation` | string | A string identifying the kind of messaging operation as defined in the [Operation names](#operation-names) section above. | `publish` | Required |
| `messaging.batch.size` | int | The number of messages sent, received, or processed in the scope of the batching operation. [1] | `0`; `1`; `2` | Conditionally Required: [2] |
| `messaging.message.conversation_id` | string | The [conversation ID](#conversations) identifying the conversation to which the message belongs, represented as a string. Sometimes called "Correlation ID". | `MyConversationId` | Recommended: [3] |

> | `messaging.message.id` | string | A value used by the messaging system as an identifier for the message, represented as a string. | `452a7c7c7c7048c2f887f61572b18fc2` | Recommended |

Author: I think we have to keep yml definition full and correct. We need to explain which per-message attributes are common (and can appear on every span), which of them apply only to publish/receive/etc.

> | `messaging.message.id` | string | A value used by the messaging system as an identifier for the message, represented as a string. | `452a7c7c7c7048c2f887f61572b18fc2` | Recommended |

Author: per-message attribute table is removed

| `messaging.message.payload_compressed_size_bytes` | int | The compressed size of the message payload in bytes. | `2048` | Recommended: [4] |
| `messaging.message.payload_size_bytes` | int | The (uncompressed) size of the message payload in bytes. Also use this attribute if it is unknown whether the compressed or uncompressed payload size is reported. | `2738` | Recommended: [5] |

> | [`net.app.protocol.name`](span-general.md) | string | Application layer protocol used. The value SHOULD be normalized to lowercase. | `AMQP`; `MQTT` | Recommended |

Reviewer 4: The value should be normalized to lowercase but then the examples are uppercase. Should we use lowercase in the examples as well? To reinforce that the should be that way.

| [`net.app.protocol.version`](span-general.md) | string | Version of the application layer protocol used. See note below. [6] | `3.1.1` | Recommended |
| [`net.peer.name`](span-general.md) | string | Logical remote hostname, see note below. [7] | `example.com` | Conditionally Required: If available. |
| [`net.sock.family`](span-general.md) | string | Protocol [address family](https://man7.org/linux/man-pages/man7/address_families.7.html) which is used for communication. | `inet6`; `bluetooth` | Conditionally Required: [8] |
| [`net.sock.peer.addr`](span-general.md) | string | Remote socket peer address: IPv4 or IPv6 for internet protocols, path for local communication, [etc](https://man7.org/linux/man-pages/man7/address_families.7.html). | `127.0.0.1`; `/tmp/mysql.sock` | Recommended |
| [`net.sock.peer.name`](span-general.md) | string | Remote socket peer name. | `proxy.example.com` | Recommended: [9] |
| [`net.sock.peer.port`](span-general.md) | int | Remote socket peer port. | `16456` | Recommended: [10] |

> **[1]:** Instrumentation SHOULD always set it on batch `receive` operations regardless of the actual number of received messages (e.g. 0, 1, or more).

Author: After discussion on SIG meeting, changed to `Instrumentations SHOULD NOT set `messaging.batch.size` on spans that operate with a single message. When client library supports batch and single-message API for the same operation, instrumentations SHOULD use `messaging.batch.size` for batching APIs and SHOULD NOT use it for single-message APIs.`

**[8]:** If different than `inet` and if any of `net.sock.peer.addr` or `net.sock.host.addr` are set. Consumers of telemetry SHOULD accept both IPv4 and IPv6 formats for the address in `net.sock.peer.addr` if `net.sock.family` is not set. This is to support instrumentations that follow previous versions of this document.

**[9]:** If different than `net.peer.name` and if `net.sock.peer.addr` is set.

**[10]:** If defined for the address family and if different than `net.peer.port` and if `net.sock.peer.addr` is set.

`messaging.operation` has the following list of well-known values. If one of them applies, then the respective value MUST be used, otherwise a custom value MAY be used.

| Value  | Description |
|---|---|
| `publish` | publish |
| `receive` | receive |
| `process` | process |
<!-- endsemconv -->

Additionally `net.peer.port` from the [network attributes][] is recommended.
Furthermore, it is strongly recommended to add the [`net.transport`][] attribute and follow its guidelines, especially for in-process queueing systems (like [Hangfire][], for example).
These attributes should be set to the broker to which the message is sent/from which it is received.

> Note that attributes in `messaging.message` namespace describe an individual message, `messaging.destination` namespace

Author: do you mean that `messaging.message.conversation_id` can be used to describe the batch? I believe with your notes above, it's the only attribute in `message` namespace that can be shared between different messages? 

I guess the point here that `message` attributes describe message properties, but as an optimization, message properties can be populated once on the corresponding span. Do you think I should change the description so it's more clear?

> Note that attributes in `messaging.message` namespace describe an individual message, `messaging.destination` namespace

Reviewer 5: This is no longer relevant after latest changes. You can resolve it (I cannot resolve it myself)

contains attributes that describe the logical entity messages are published to, and `messaging.source` describes
logical entity messages are received from.

[network attributes]: span-general.md#general-network-connection-attributes
[`net.transport`]: span-general.md#network-transport-attributes
[Hangfire]: https://www.hangfire.io/

### Producer attributes

> For message producers, the following additional attributes may be set:

Reviewer 5: Here we say "**may** be set", but it's a bit confusing, as the required level for these attributes is not MAY.
They are not optional, they should be set if available.
Maybe replace the word may with something else?

| `messaging.destination.name` | string | The message destination name [6] | `MyQueue`; `MyTopic` | Conditionally Required: [7] |

**[1]:** If the message destination is either a `queue` or `topic`.

**[2]:** Destination names could be constructed from templates. An example would be a destination name involving a user name or product id. Although the destination name in this case is of high cardinality, the underlying template is of low cardinality and can be effectively used for grouping and searching spans.

**[3]:** If available. Instrumentations MUST NOT use `messaging.destination.name` as template unless low-cardinality of destination name is guaranteed.

**[4]:** If value is `true`. When missing, the value is assumed to be `false`.

**[5]:** If value is `true`. When missing, the value is assumed to be `false`.

**[6]:** Destination name SHOULD uniquely identify specific queue, topic, or other entity within broker. If
broker does not have such notion, destination name SHOULD uniquely identify broker.

> **[7]:** If the value applies to all messages in the batch.

Reviewer 5: I know it might be a bit too specific, but for a batch of messages the value can be the same by protocol (reading a batch of messages from a single queue), or it can be that we read a few arbitrary messages and they happen to have the same destination at a single invocation.
I think this attribute is required for the first case (same destination by definition), and SHOULD NOT be set for the second case.
WDYT?

> **[7]:** If the value applies to all messages in the batch.

Author: > I know it might be a bit too specific, but for a batch of messages the value can be the same by protocol (reading a batch of messages from a single queue), or it can be that we read a few arbitrary messages and they happen to have the same destination at a single invocation.
I think this attribute is required for the first case (same destination by definition), and SHOULD NOT be set for the second case.

why SHOULD NOT? What if I'm sending 100 messages to a single destination on broker that supports per-message destination - do I pay more for this attribute repeated 100 times?  Probably not. I think instrumentation should do its best effort to avoid redundancy.

`messaging.destination.kind` MUST be one of the following:

| Value  | Description |
|---|---|
| `queue` | A message sent to a queue |
| `topic` | A message sent to a topic |
<!-- endsemconv -->

### Consumer attributes

For message consumers, the following additional attributes may be set:

<!-- semconv messaging.consumer -->
| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|
| `messaging.consumer_id` | string | The identifier for the consumer receiving a message. For Kafka, set it to `{messaging.kafka.consumer_group} - {messaging.kafka.client_id}`, if both are present, or only `messaging.kafka.consumer_group`. For brokers, such as RabbitMQ and Artemis, set it to the `client_id` of the client consuming the message. | `mygroup - client-6` | Recommended |
| `messaging.source.kind` | string | The kind of message source | `queue` | Conditionally Required: [1] |

> | `messaging.source.template` | string | Low cardinality field representing messaging source | `/customers/{customerId}` | Conditionally Required: [2] |

Reviewer 5: for destination, we have this text describing what it means:

> Destination names could be constructed from templates. An example would be a destination name involving a user name or product id. Although the destination name in this case is of high cardinality, the underlying template is of low cardinality and can be effectively used for grouping and searching spans.

Maybe add it here so a user that landed on this section has the context as well? or extract to a common section and link both to the description?

| `messaging.source.temporary` | boolean | denotes that the source is a temporary source and might not exist anymore after messages are processed |  | Recommended: [3] |

> | `messaging.source.anonymous` | boolean | denotes that the source is an anonymous source (could be unnamed or have auto-generated name) |  | Recommended: [4] |

Reviewer 5: for destination we wrote:

> A boolean that is true if the message destination is temporary and might not exist anymore after messages are processed.

maybe we should align both descriptions for consistency?

Even though in that case one might think that the processing span's kind should be `INTERNAL`, that kind MUST NOT be used.
Instead span kind should be set to either `CONSUMER` or `SERVER` according to the rules defined above.

### Per-message attributes

> Note that each or messaging operations (`send`, `receive`, or `process`) can describe a batch of messages. For batch operations per-message attributes cannot be set on

Author: Let's discuss it tomorrow. It'd be great to clarify recommendations for the non-batching case.  I think it'd be overkill to use links for rabbit or other single-send systems. Also requiring attributes on links means that this spec will stop working for all backends.

> Note that each or messaging operations (`send`, `receive`, or `process`) can describe a batch of messages. For batch operations per-message attributes cannot be set on

Reviewer 2: I will add to the agenda for today's meeting.

> Note that each or messaging operations (`send`, `receive`, or `process`) can describe a batch of messages. For batch operations per-message attributes cannot be set on

Author: discussed on SIG meeting: don't be prescriptive for now, clarify links in future PRs along with link/event choice

> corresponding span and SHOULD instead be set on link. See [Batch Receiving](#batch-receiving) and [Batch Processing](#batch-processing) for more information on correlation using links.

Reviewer 5: It is implied from the SHOULD wording, but I wonder if it is appropriate to document here that instrumentation is allowed to not create some or all links for reasons of volumes/costs or if the information is not available

For example, `message.id` is not always defined for all systems (redis pub-sub, sockeit.io).
Some attributes can be present but supply no value to the user in some cases - `messaging.message.payload_size_bytes` can be recorded but it can be considered noise for many users.
Maybe some attributes are simply too expensive or complicated to capture.

If instrumentation doesn't need or chooses not to store anything on links attributes, they may be dropped entirely IMO.

Author: all of the message attributes come with `recommended` requirement level which [explicitly](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-requirement-level.md#recommended) mentions they could be disabled.

> corresponding span and SHOULD instead be set on link. See [Batch Receiving](#batch-receiving) and [Batch Processing](#batch-processing) for more information on correlation using links.

Reviewer 5: Creating a link requires a context to set as the link target. This context might not always be available:
- producer is not instrumented
- no appropriate mechanism to propagate context in a specific system (redis pub-sub, socket.io for example)
- intermediaries and message routing mechanisms which drop the context.

However, using links is the only way to record per message information in batch operations, when might be desired even when context is not available.

Do you think we should also describe how to generate links in the absence of creation context? we discussed using an invalid (zeroed) context for that if I remember correctly. We can also add it in a future PR but I think it should eventually be speced.

Reviewer 6: > Do you think we should also describe how to generate links in the absence of creation context?

I don't think links are ideal then, maybe use events instead?

> Do you think we should also describe how to generate links in the absence of creation context? we discussed using an invalid (zeroed) context for that if I remember correctly.

I don't think SDKs (consistently?) support creating invalid links right now (tangentially related #753).

And also: How should this be handled on the sending side? A link (necessarily) always points to a parent, so I don't think it even makes sense semantically. This feels like a very important issue/inconsistency to me. Should span events be used instead?

Reviewer 5: The problem with span events is that they don't capture a context, thus making it impossible to correlate a single message flow when producer and consumer are not in the same trace (or even when they are in the same trace sometimes?)

Links offer both the silo to store per-message attributes, and a way to "jump" from consumer to producer.

Reviewer 6: I was talking about using events for where you don't have a context you can capture.

Reviewer 6: I agree, it's not nice at all. And events don't really make sense semantically either, but IMHO more than links on the outgoing side

Author: I want to refrain from changing the current state of affairs in this PR. Current spec already uses links and not events and does not deal with the edge cases (missing context on the message). We'll nee to solve it in scope of all messaging work. but let's not do it in this PR and keep status quo for now.

Author: I ended up changing this to 

>  For batch operations per-message attributes cannot be set on the corresponding span and MAY instead be set on a link

this way events can also be used. 

Once we finalize the consensus in messaging SIG on span structure, we'll bring it here via different PR.

Reviewer 6: > Current spec already uses links and not events and does not deal with the edge cases (missing context on the message).

The current spec doesn't use attributes on links, so I always interpreted that as creating links only for the messages you have context.
I don't think an incoming message without context is an edge case, it seems perfectly normal to me.

Author: > The current spec doesn't use attributes on links, so I always interpreted that as creating links only for messages with context.

Right, but the current spec already uses links (and not events). The only thing I want to make with this change is to allow per-message attributes to appear on links.
This change makes current spec limitations more obvious but does not change the status quo.

We'll have another PR on when and how links are populated separately. We can discuss events/links/empty context there

> corresponding span and SHOULD instead be set on link. See [Batch Receiving](#batch-receiving) and [Batch Processing](#batch-processing) for more information on correlation using links.

Reviewer 5: I think it is also appropriate to state here that each message should be captured as link only once and not more.
In the absence of clear systematic rules of how to differentiate specific messages by attributes, and the fact that links can be dropped, It will not be possible to tell if a span with `messaging.batch.size` of 2 and two links contains one link per message, or zero link for the first message and two for the seconds.

Constraining the messages to be recorded at most once as links, will allow UIs to derive a meaningful set of unique messages from links with no ambiguaty.

> Following attributes apply to links describing each message.

Reviewer 2: Somewhat related but maybe not: Should we remove the message-specific attributes from the general table and only have them here? Not sure but it might confuse people on why they are duplicated.

Even though in that case one might think that the processing span's kind should be `INTERNAL`, that kind MUST NOT be used.
Instead span kind should be set to either `CONSUMER` or `SERVER` according to the rules defined above.

### Per-message attributes

All messaging operations (`send`, `receive`, `process`, or other not covered by this specification) can describe a batch of messages. For batch operations per-message attributes cannot be set on the corresponding span and MAY instead be set on a link. See [Batch Receiving](#batch-receiving) and [Batch Processing](#batch-processing) for more information on correlation using links.

> Following attributes apply to links describing each message.

Reviewer 2: ```suggestion
The following attributes apply to links describing each message on a batch.
```
We want to say the following attributes apply to links when working with a batch right? Or regardless? I think we didn't reach a decision if we will _always_ record message-specific attributes on a link, regardless if it's a batch or not. Maybe we should pin this down, so we know how to phrase things here.

Author: We discussed attributes on span vs links on Thu meeting and decided to not be prescriptive in this PR just yet. So I rephrased it, LMK what you think

Even though in that case one might think that the processing span's kind should be `INTERNAL`, that kind MUST NOT be used.
Instead span kind should be set to either `CONSUMER` or `SERVER` according to the rules defined above.

### Per-message attributes

All messaging operations (`publish`, `receive`, `process`, or other not covered by this specification) can describe a batch of messages.

> The following attributes describe individual messages. For batch operations per-message attributes cannot be set on the corresponding unless they are the same for all batch and MAY instead be set on a link. See [Batch Receiving](#batch-receiving) and [Batch Processing](#batch-processing) for more information on correlation using links.

Reviewer 5: ```suggestion
The following attributes describe individual messages. For batch operations per-message attributes cannot be set on the corresponding span attributes unless they are the same for all messages in the batch and MAY instead be set on a link. See [Batch Receiving](#batch-receiving) and [Batch Processing](#batch-processing) for more information on correlation using links.
```

| `messaging.message.id` | string | A value used by the messaging system as an identifier for the message, represented as a string. | `452a7c7c7c7048c2f887f61572b18fc2` | Recommended |
| `messaging.message.conversation_id` | string | The [conversation ID](#conversations) identifying the conversation to which the message belongs, represented as a string. Sometimes called "Correlation ID". | `MyConversationId` | Recommended |
| `messaging.message.payload_size_bytes` | int | The (uncompressed) size of the message payload in bytes. Also use this attribute if it is unknown whether the compressed or uncompressed payload size is reported. | `2738` | Recommended |
| `messaging.message.payload_compressed_size_bytes` | int | The compressed size of the message payload in bytes. | `2048` | Recommended |

**[1]:** Destination name SHOULD uniquely identify specific queue, topic, or other entity within broker. If
broker does not have such notion, destination name SHOULD uniquely identify broker.

**[2]:** Source name SHOULD uniquely identify specific queue, topic, or other entity within broker. If
broker does not have such notion, source name SHOULD uniquely identify broker.

> <!-- endsemconv -->

Reviewer 5: Do we have to repeat the table here? 
Seems a bit confusing, as these attributes are now specified twice in the document.

Also, some attributes never makes sense in a batch or need clarification on their meaning for batch (`messaging.message.id`, `messaging.message.payload_size_bytes`, `messaging.message.payload_compressed_size_bytes`)

Author: If we want yaml to be accurate and strict, we need to specify per-message attributes for each span. We also need to specify attributes that can go on links.

### Attributes specific to certain messaging systems

> All attributes that are specific for a messaging system SHOULD be populated in `messaging.{system}` namespace. Message-, destination-, or source-specific attributes SHOULD be populated under corresponding namespace:

Reviewer 5: What is the meaning of the hyphen? `Message-`, `destination-`?

Reviewer 3: ```suggestion
All attributes that are specific for a messaging system SHOULD be populated in `messaging.{system}` namespace. Attributes that are specific to the message, the message destination, or the message source SHOULD be populated under corresponding namespace:
```

* Message-specific under `messaging.{system}.message`
* Destination-specific under `messaging.{system}.destination`

> * source-specific under `messaging.{system}.source`

Reviewer 5: These sentences are not consistent with the capital first letter.
Since they are found in attribute names in lower case, I suggest making them all lowercase.

#### RabbitMQ

> In RabbitMQ, the destination is defined by an *exchange* and a *routing key*.

Reviewer 3: In the line below, `messaging.destination` should be changed to `messaging.destination.name`.

Some messaging systems such as Kafka, Azure EventGrid allow to publish a single batch of messages to different topics, in such cases attributes in `messaging.destination` and `messaging.source` MAY be
set on links. Corresponding instrumentations MAY set source and destination attributes on the span if all messages in the batch share the same destination or source.

### Attributes specific to certain messaging systems

All attributes that are specific for a messaging system SHOULD be populated in `messaging.{system}` namespace. Attributes that are specific to the message, the message destination, or the message source SHOULD be populated under corresponding namespace:

* message-specific under `messaging.{system}.message`
* destination-specific under `messaging.{system}.destination`
* source-specific under `messaging.{system}.source`

#### RabbitMQ

In RabbitMQ, the destination is defined by an *exchange* and a *routing key*.
`messaging.destination.name` MUST be set to the name of the exchange. This will be an empty string if the default exchange is used.

<!-- semconv messaging.rabbitmq -->
| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|

> | `messaging.rabbitmq.message.routing_key` | string | RabbitMQ message routing key. | `myKey` | Conditionally Required: If not empty. |

Author: Per Rabbit docs, [routing key is a message property](https://www.rabbitmq.com/tutorials/amqp-concepts.html#messages). It's [also available](https://www.rabbitmq.com/consumers.html#message-properties) on consumers

<!-- semconv messaging.kafka -->
| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|
| `messaging.kafka.message.key` | string | Message keys in Kafka are used for grouping alike messages to ensure they're processed on the same partition. They differ from `messaging.message.id` in that they're not unique. If the key is `null`, the attribute MUST NOT be set. [1] | `myKey` | Recommended |
| `messaging.kafka.consumer_group` | string | Name of the Kafka Consumer Group that is handling the message. Only applies to consumers, not producers. | `my-group` | Recommended |
| `messaging.kafka.client_id` | string | Client Id for the Consumer or Producer that is handling the message. | `client-5` | Recommended |

> | `messaging.kafka.message.partition` | int | Partition the message is sent to. | `2` | Recommended |

Reviewer 3: I think it's right to have the partition as part of the `message` namespace. In Kafka, you can have message from different partitions in a single batch (see [here](https://stackoverflow.com/questions/64423092/consuming-kafka-batch-for-multiple-partitions)). This can't be modelled if the partition is part of the `source` or `destination` namespace.

> | `messaging.kafka.message.partition` | int | Partition the message is sent to. | `2` | Recommended |

Author: agreed with @pyohannes  - partition is message property ([ProducerRecord](https://kafka.apache.org/24/javadoc/index.html?org/apache/kafka/clients/producer/ProducerRecord.html), [ConsumerRecord](https://kafka.apache.org/24/javadoc/index.html?org/apache/kafka/clients/consumer/ConsumerRecord.html))

<!-- semconv messaging.kafka -->
| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|
| `messaging.kafka.message.key` | string | Message keys in Kafka are used for grouping alike messages to ensure they're processed on the same partition. They differ from `messaging.message.id` in that they're not unique. If the key is `null`, the attribute MUST NOT be set. [1] | `myKey` | Recommended |

> | `messaging.kafka.consumer_group` | string | Name of the Kafka Consumer Group that is handling the message. Only applies to consumers, not producers. | `my-group` | Recommended |

Reviewer 5: Should this be changed to `messaging.kafka.source.consumer_group`?

Author: consumer group is a group of consumers, it does no describe source

| `messaging.kafka.client_id` | string | Client Id for the Consumer or Producer that is handling the message. | `client-5` | Recommended |

> | `messaging.kafka.partition` | int | Partition the message is sent to. | `2` | Recommended |

Reviewer 5: I wonder if the partition should be hosted under a namespace 🤔 (`source`, `destination`)

Author: moved both to `messaging.kafka.message` namespace and also did the same for `messaging.rabbitmq.message.routing_key` as it's message-specific

|-|-|-|-|-|-|
| Span name | `"Q send"` | `"Q send"` | `"Q receive"` | `"Q receive"` | `"Q process"` |
| Parent |  |  | Span Prod1 | Span Prod2 |  |
| Links |  |  |  |  | [Span Prod1, Span Prod2 ] |

> | Link attributes |  |  |  |  | Span Prod1: `messaging.message.id`: `"a1"`  |

Author: I'm as surprised as you, but current spec  does not use links on receive spans. The idea it has is:
- Receive  batch has ho links
- Receive process spans are its children and there is process span per child
- Process spans have links to message context

I.e. receive is still correlated to message, but via link on the child.

> | Link attributes |  |  |  |  | Span Prod1: `messaging.message.id`: `"a1"`  |

Reviewer 3: The idea to put message specific attributes on links works for the new span structure we worked out in the messaging workgroup, however, it will not fit with some of the examples in this spec (as there are cases like this one, where there are no links to hold attributes).

> | `messaging.destination` | `"Q"` | `"Q"` | `"Q"` | `"Q"` | `"Q"` |

Reviewer 3: We probably also need to fix attribute names throughout the examples (like changing this line to `messaging.destination.name`).
